packages/drivers/odsp-driver/src/snapshotFormatConverter.ts

```diff
@@ -0,0 +1,279 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert, stringToBuffer } from "@fluidframework/common-utils";
+import { IBlob, ISnapshotTree } from "@fluidframework/protocol-definitions";
+import { ISequencedDeltaOpMessage } from "./contracts";
+import { BlobCore, NodeCore, ReadBuffer, TreeBuilder } from "./snapshotRepresentation";
+
+const minReadVersion = "1.0";
+
+/**
+ *  Header of the downloaded snapshot.
+ */
+export interface ISnapshotHeader {
+    // This is the minimum version of the reader required to read the wire format of a given snapshot.
+    minReadVersion: string,
+    // Represents the version with which snapshot is created.
+    createVersion: string,
+    // Seq number at which the snapshot is created.
+    snapshotSeqNumber: number,
+}
+
+/**
+ * Class used to convert snapshot formats. It can convert binary compact representation to
+ * odsp snapshot and vice-versa.
+ */
+export class SnapshotFormatConverter {
+    constructor() {}
+
+    /**
+     * Recreates header section and then validate that.
+     * @param node - tree node to read header section from
+     */
+    private readAndValidateHeaderSection(node: NodeCore) {
+        assert(node.getString(0) === "MinReadVersion", "MinReadVersion header should be present");
+        assert(node.getString(2) === "CreateVersion", "CreateVersion header should be present");
+        assert(node.getString(4) === "SnapshotSequenceNumber", "SnapshotSequenceNumber header should be present");
+        const header = {
+            minReadVersion: node.getString(1),
+            createVersion: node.getString(3),
+            snapshotSeqNumber: node.getNumber(5),
+        };
+        assert(minReadVersion >= header.minReadVersion, "Driver min read version should >= to server minReadVersion");
```

Comment:
We should consider version range that client understands being communicated to server as part of request, such that server could support different versions. This does not change the need on client to validate that client can understand output

Reply:
Hmm. Will talk to server folks about it.

Reply:
If we are going with semantic versioning, an idea might be give us the semantic major version that you would like to read and we will return the newest version that's within that major version? like: give us 2 and we will return the newest version that's at least 2.0.0 but no more than 3.0.0, this way it doesn't break the client reading and it uses only one version instead of a version range. If you need more specific version, just specify the minor version, if all three versions are specified, we will give you that exact version.

Reply:
This sounds great @sunnyhuang2008 . I think we can proceed with this.